src/lib.rs

```diff
@@ -150,110 +213,260 @@ impl<'object, Endian> DebugInfo<'object, Endian>
+                        // current is better
+                        Some((prev, prange, pdist))
+                    };
                 } else {
-                    current = Some(row.clone());
+                    func = Some((p, range, dist));
                 }
+
+                // Unfortunately, we're not done even if we've already found loc, because there
+                // *may* be another subprogram (specifically, one that has been inlined) that
+                // matches us better. We need to keep going until we find the best one.
             }
+
+            return Some((path, line, func.map(|u| u.0.name.to_string_lossy())));
         }
         None
     }
 }
 
 // TODO: most of this should be moved to the main library.
-struct Unit<'input> {
+use std::marker::PhantomData;
+struct Unit<'input, Endian> {
     address_size: u8,
     ranges: Vec<gimli::Range>,
     line_offset: gimli::DebugLineOffset,
     comp_dir: Option<&'input std::ffi::CStr>,
     comp_name: Option<&'input std::ffi::CStr>,
+    programs: Vec<Program<'input>>,
+    phantom: PhantomData<Endian>,
 }
 
-impl<'input> Unit<'input> {
-    fn parse<Endian>(debug_abbrev: &gimli::DebugAbbrev<Endian>,
-                     debug_ranges: &gimli::DebugRanges<Endian>,
-                     debug_str: &gimli::DebugStr<'input, Endian>,
-                     header: &gimli::CompilationUnitHeader<'input, Endian>)
-                     -> Option<Unit<'input>>
-        where Endian: gimli::Endianity
-    {
+impl<'input, Endian> Unit<'input, Endian>
+    where Endian: gimli::Endianity
+{
+    fn parse(debug_abbrev: &gimli::DebugAbbrev<Endian>,
+             debug_ranges: &gimli::DebugRanges<Endian>,
+             debug_str: &gimli::DebugStr<'input, Endian>,
+             header: &gimli::CompilationUnitHeader<'input, Endian>,
+             func_names: Option<&HashMap<usize, &'input std::ffi::CStr>>)
+             -> Option<Unit<'input, Endian>> {
+
+        // We first want to parse out the compilation unit, and then any contained subprograms.
         let abbrev = header.abbreviations(*debug_abbrev).expect("Fail");
         let mut entries = header.entries(&abbrev);
-        let (_, entry) = entries.next_dfs()
-            .expect("Should parse first entry OK")
-            .expect("And first entry should exist!");
-        assert_eq!(entry.tag(), gimli::DW_TAG_compile_unit);
-
-        let ranges = if let Some(ranges) =
-            Self::parse_noncontiguous_ranges(entry, debug_ranges, header.address_size()) {
-            ranges
-        } else if let Some(range) = Self::parse_contiguous_range(entry) {
-            vec![range]
-        } else {
-            return None;
-        };
+        let mut unit = {
+            // Scoped so that we can continue using entries for the loop below
+            let first = entries.next_dfs();
+            let (_, entry) = first.expect("Should parse first entry OK")
+                .expect("And first entry should exist!");
```

Comment:
Now that this is a library, rather than an example program, we probably shouldn't panic on malformed DWARF, so we should probably early return `None` here rather than `expect("...")`ing that the input data is valid.